opengever/maintenance/scripts/archival_file_checker.py

```diff
@@ -0,0 +1,394 @@
+"""
+Script that reports documents missing their archival file, and optionally
+queues them for conversion by a nightly job.
+
+Usage: archival_file_checker.py [-n] [report_missing | queue_missing]
+
+The script takes one of two commmands:
+
+report_missing
+    Reports all documents missing their archival file.
+
+queue_missing
+    Reports the documents, and additionally queues them for conversion by
+    a nightly job (by adding them to a persistent queue on the site root).
+
+The script will display some basic console output, and automatically log
+that ouput and some more detailed information to a logfile.
+"""
+from App.config import getConfiguration
+from BTrees.IIBTree import IITreeSet
+from BTrees.IOBTree import IOBTree
+from collections import Counter
+from collections import OrderedDict
+from datetime import datetime
+from ftw.bumblebee.interfaces import IBumblebeeDocument
+from opengever.document.behaviors import IBaseDocument
+from opengever.dossier.behaviors.dossier import IDossierMarker
+from opengever.maintenance.debughelpers import setup_app
+from opengever.maintenance.debughelpers import setup_plone
+from opengever.maintenance.nightly_archival_file_job import MISSING_ARCHIVAL_FILE_KEY
+from opengever.maintenance.utils import TextTable
+from plone import api
+from Products.CMFPlone.interfaces import IPloneSiteRoot
+from zope.annotation import IAnnotations
+from zope.component import getUtility
+from zope.intid.interfaces import IIntIds
+import argparse
+import logging
+import os
+import sys
+import transaction
+
+
+logger = logging.getLogger('archival_file_checker')
+logger.setLevel(logging.INFO)
+
+
+class ArchivalFileChecker(object):
+    """Checks whether documents that should have an archival file actually
+    do have one, and reports the ones that don't.
+
+    With command 'queue_missing' also queues those documents for conversion
+    via a nightly job.
+    """
+
+    def __init__(self, context, options, logger):
+        self.context = context
+        self.options = options
+        self.logger = logger
+
+        self.log = logger.log
+        self.log_to_file = logger.log_to_file
+
+        self.catalog = api.portal.get_tool('portal_catalog')
+        self.all_dossier_stats = None
+
+    def run(self):
+        missing_by_dossier = self.check()
```

Comment:
Basically `run` will fail if the context is not the site root. But it will fail only after doing all the work. I'd add the check `assert IPloneSiteRoot.providedBy(self.context)` wither here before running the `self.check()` or even in the `__init__`